fl4health/clients/apfl_client.py

```diff
@@ -1,255 +1,101 @@
-from fl4health.utils.metrics import Metric, MetricAverageMeter, MetricMeter
+from fl4health.parameter_exchange.layer_exchanger import FixedLayerExchanger
+from fl4health.utils.losses import Losses, LossMeterType
+from fl4health.utils.metrics import Metric, MetricMeterType
 
-LocalLoss = torch.Tensor
-GlobalLoss = torch.Tensor
-PersonalLoss = torch.Tensor
 
-LocalPreds = torch.Tensor
-GlobalPreds = torch.Tensor
-PersonalPreds = torch.Tensor
-
-ApflTrainStepOutputs = Tuple[LocalLoss, GlobalLoss, PersonalLoss, LocalPreds, GlobalPreds, PersonalPreds]
-
-
-class ApflClient(NumpyFlClient):
+class ApflClient(BasicClient):
     def __init__(
         self,
         data_path: Path,
         metrics: Sequence[Metric],
         device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+        use_wandb_reporter: bool = False,
+        checkpointer: Optional[TorchCheckpointer] = None,
     ) -> None:
-        super().__init__(data_path, device)
-        self.metrics = metrics
-        self.model: APFLModule
-        self.train_loader: DataLoader
-        self.val_loader: DataLoader
-        self.num_examples: Dict[str, int]
-        self.criterion: _Loss
-        self.local_optimizer: torch.optim.Optimizer
-        self.global_optimizer: torch.optim.Optimizer
-
-    def is_start_of_local_training(self, epoch: int, step: int) -> bool:
-        return epoch == 0 and step == 0
-
-    def fit(self, parameters: NDArrays, config: Config) -> Tuple[NDArrays, int, Dict[str, Scalar]]:
-        if not self.initialized:
-            self.setup_client(config)
-
-        self.set_parameters(parameters, config)
-        local_epochs = self.narrow_config_type(config, "local_epochs", int)
-
-        # Default APFL uses an average meter
-        global_meter = MetricAverageMeter(self.metrics, "global")
-        local_meter = MetricAverageMeter(self.metrics, "local")
-        personal_meter = MetricAverageMeter(self.metrics, "personal")
-        # By default the APFL client trains by epochs
-        metric_values = self.train_by_epochs(local_epochs, global_meter, local_meter, personal_meter)
-        # FitRes should contain local parameters, number of examples on client, and a dictionary holding metrics
-        # calculation results.
-        return (
-            self.get_parameters(config),
-            self.num_examples["train_set"],
-            metric_values,
-        )
-
-    def evaluate(self, parameters: NDArrays, config: Config) -> Tuple[float, int, Dict[str, Scalar]]:
-        if not self.initialized:
-            self.setup_client(config)
-
-        self.set_parameters(parameters, config)
-        # Default APFL uses an average meter
-        global_meter = MetricAverageMeter(self.metrics, "global")
-        local_meter = MetricAverageMeter(self.metrics, "local")
-        personal_meter = MetricAverageMeter(self.metrics, "personal")
```

Comment:
Losing these individual meters is a bit of a shame. As stated for the global, local, and personal losses below, these meters helped me understand what was happening for APFL under the hood, so to speak, during training (i.e. whether it made sense that alpha was moving one way or another).

Could we define a `Meters` class, similar to the Losses class, but that contains Meter objects and work with that in the BasicClient?

Reply:
Updated in https://github.com/VectorInstitute/FL4Health/pull/62/commits/f462d69c915f7f19a2d97d808e685d1d0ad39dd6